ADR 001 ABCI++ (#231)
celestiaorg/celestia-app: docs/architecture/ADR-001-ABCI++.md

# ADR 001: ABCI++ Adoption

## Changelog

- 2022-03-03: Initial Commit

## Context

Among other things, ABCI++ enables arbitrary logic to be performed by the application to create and verify proposal blocks.

We need this functionality in order for block producers to:

- Pick a square size / fill the data square efficiently
- Malleate user transactions
- Separate rollup block data from the stateful portion of the transaction
- Use the appropriate signature for the selected block size
- Follow the [non-interactive default rules](https://github.com/celestiaorg/celestia-specs/blob/65d4d528beafaf8d84f7eb805db940f05f2a2c93/src/rationale/message_block_layout.md#non-interactive-default-rules) (not done here)
- Create Celestia-specific data hash by erasure-coding the block data

We also need this functionality for validators to verify that:

> - For every `MsgPayForMessage` included in the block, there is also a corresponding message and vice versa.

Reviewer: ```suggestion
- For every `MsgPayForData` (previously `MsgPayForMessage`) included in the block, there is also a corresponding message and vice versa.
```

type Application interface {
   ...
   PrepareProposal(RequestPrepareProposal) ResponsePrepareProposal
   ProcessProposal(RequestProcessProposal) ResponseProcessProposal
   ...
}
```

It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modified to pass the entire block data. This is because Celestia separates some block data that cannot modify state (messages), and the application has to have access to both normal transaction data and messages to perform the necessary processing and checks.
```protobuf
message RequestPrepareProposal {
 // block_data is an array of transactions that will be included in a block,
 // sent to the app for possible modifications.
 // applications can not exceed the size of the data passed to it.
 tendermint.types.Data block_data = 1;
 // If an application decides to populate block_data with extra information, they can not exceed this value.
 int64 block_data_size = 2;
}

message RequestProcessProposal {
 tendermint.types.Header header     = 1 [(gogoproto.nullable) = false];
 tendermint.types.Data   block_data = 2;
}

...

> // Data contains the set of transactions, evidence, intermediate state roots,

Reviewer: ```suggestion
// Data contains the set of transactions, evidence,
```

// and messages to be included in the block
message Data {
 // Txs that will be applied by state @ block.Height+1.
 // NOTE: not all txs here are valid.  We're just agreeing on the order first.
 // This means that block.AppHash does not include these txs.
 repeated bytes txs = 1;

>  IntermediateStateRoots intermediate_state_roots = 2 [(gogoproto.nullable) = false];

Reviewer: I'd already skip it here tbh.
Add a comment that field number to is reserved for ISRs instead.

Author: good idea :+1: https://github.com/celestiaorg/celestia-app/pull/231/commits/9dace036b5c51076f73e8c38699b5360b11ee3a1

   if err != nil {
       return false, err
   }

   if !sqwr.hasRoomForTx(delimTx) {
       return false, nil
   }

   sqwr.txWriter.Write(delimTx)
   return true, nil
}

// writeMalleatedTx malleates a MsgWirePayForMessage into a MsgPayForMessage and
// its corresponding message provided that it has a MsgPayForMessage for the
// preselected square size. Returns true if the write was successful, false if
// there was not enough room in the square.
func (sqwr *squareWriter) writeMalleatedTx(
   parentHash []byte,
   tx signing.Tx,
   wpfm *types.MsgWirePayForMessage,
) (ok bool, malleatedTx coretypes.Tx, msg *core.Message, err error) {
   ... // process the malleated tx and extract the message.

   // check if we have room for both the tx and message it is crucial that we
   // add both atomically, otherwise the block is invalid
   if !sqwr.hasRoomForBoth(wrappedTx, coreMsg.Data) {
       return false, nil, nil, nil
   }

   ...

   sqwr.txWriter.Write(delimTx)
   sqwr.msgWriter.Write(coretypes.Message{
       NamespaceID: coreMsg.NamespaceId,
       Data:        coreMsg.Data,
   })

   return true, wrappedTx, coreMsg, nil
}
```

> Lastly, the data availability header is used to create the `DataHash` in the `Header` in the application instead of in tendermint. This is done by modifying the protobuf version of the block data to retain the cached hash and setting it during `ProcessProposal`. Later, in `ProcessProposal` other fullnodes check that the `DataHash` matches the block data by recomputing it. Previously, this extra check was performed inside the `ValidateBasic` method of `types.Data`, where is was computed each time it was decoded. Not only is this more efficient as it saves significant computational resources and keeps `ValidateBasic` light, it is also much more explicit. This approach does not however dramatically any existing code in tendermint, as the code to compute the hash of the block data remains there. Ideally, we would move all of the code that computes erasure encoding to the app. This approach allows us to keep the intuitiveness of the `Hash` method for `types.Data`, along with not forcing us to change many tests in tendermint, which rely on this functionality.

Reviewer: This approach does not however dramatically **change** any existing code in tendermint
```suggestion
Lastly, the data availability header is used to create the `DataHash` in the `Header` in the application instead of in tendermint. This is done by modifying the protobuf version of the block data to retain the cached hash and setting it during `ProcessProposal`. Later, in `ProcessProposal` other fullnodes check that the `DataHash` matches the block data by recomputing it. Previously, this extra check was performed inside the `ValidateBasic` method of `types.Data`, where is was computed each time it was decoded. Not only is this more efficient as it saves significant computational resources and keeps `ValidateBasic` light, it is also much more explicit. This approach does not however dramatically change any existing code in tendermint, as the code to compute the hash of the block data remains there. Ideally, we would move all of the code that computes erasure encoding to the app. This approach allows us to keep the intuitiveness of the `Hash` method for `types.Data`, along with not forcing us to change many tests in tendermint, which rely on this functionality.
```

# ADR 001: ABCI++ Adoption

## Changelog

- 2022-03-03: Initial Commit

## Context

Among other things, ABCI++ enables arbitrary logic to be performed by the application to create and verify proposal blocks.

We need this functionality in order for block producers to

> - Pick a square size / Fill the data square efficiently

Reviewer 2: ```suggestion
- Pick a square size / fill the data square efficiently
```

- Malleate user transactions
- Separate rollup block data from the stateful portion of the transaction
- Use the appropriate signature for the selected block size

> - Follow the non-interactive default rules (not done here)

Author: good idea 809e46e03f841dfaeea3b84ea8e2956cb4365157

> - Follow the [non-interactive default rules](https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md#non-interactive-default-rules) (not done here)

Reviewer: Use perma-link:
```suggestion
- Follow the [non-interactive default rules](https://github.com/celestiaorg/celestia-specs/blob/65d4d528beafaf8d84f7eb805db940f05f2a2c93/src/rationale/message_block_layout.md#non-interactive-default-rules) (not done here)
```

- Create celestia specific data hash by erasuring the block data

We also need this functionality for validators to verify that

- For every `MsgPayForMessage` included in the block, there is also a corresponding message and vice versa.

> - That the data hash represents the properly erasured block data for the selected block size

Reviewer 2: ```suggestion
- That the data hash represents the properly-erasure-coded block data for the selected block size.
```

## Alternative Approaches

While the adoption of ABCI++ is inevitable given the already made decision by upstream to implement it, here are some alternatives to the features that we need that do not use ABCI++.

Alternatives for Message Inclusion
https://github.com/celestiaorg/celestia-app/blob/92341dd68ee6e555ec6c0bb780afa3a1c8243a93/adrs/adr008:adopt-ABC%2B%2B-early.md#alternative-approaches

Alternatives for Picking a square size
https://github.com/celestiaorg/celestia-core/issues/454

## Decision

Proposed and initial implementation is complete.

## Detailed Design

### [#631](https://github.com/celestiaorg/celestia-core/pull/631) Simplified version of ABCI++ (`celestia-core`)

Here we are adding only the two new methods that are necessary for the features that we need.  
```go
// Application is an interface that enables any finite, deterministic state machine
// to be driven by a blockchain-based replication engine via the ABCI.
type Application interface {
   ...
   PrepareProposal(RequestPrepareProposal) ResponsePrepareProposal
   ProcessProposal(RequestProcessProposal) ResponseProcessProposal
   ...
}
```

> It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modifed to pass the entire block data.

Reviewer 3: ```suggestion
It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modified to pass the entire block data.
```

> It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modified to pass the entire block data.

Author: we did not, but I think its a net negative for most projects. Projects that need more will still have to fork, and it saves the majority of projects the extra step to access the txs, as the only type in upstream's `types.Data` is `[]types.Tx`.

> It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modified to pass the entire block data.

Reviewer: I think would be really cool if there was an extra field in Data for purposes like this. On the other hand, txs can already be anything and could be used for whatever the app developer has in mind. Resolving for now.

> It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modified to pass the entire block data.

Reviewer: One nit: can this ADR also mention that Block.Data here is slightly different from vanilla tendermint (one sentence maybe stating that we also need Evidence be part of the available data). I want the ADR to be straightforward to understand even from someone coming from vanilla tendermint and trying to see how we use abci++.

Author: great point https://github.com/celestiaorg/celestia-app/pull/231/commits/c9cb6063c939ac82b0a999f01f80292e6c19a2d5

```protobuf
message RequestPrepareProposal {
 // block_data is an array of transactions that will be included in a block,
 // sent to the app for possible modifications.
 // applications can not exceed the size of the data passed to it.
 tendermint.types.Data block_data = 1;
 // If an application decides to populate block_data with extra information, they can not exceed this value.
 int64 block_data_size = 2;
}

message RequestProcessProposal {
 tendermint.types.Header header     = 1 [(gogoproto.nullable) = false];
 tendermint.types.Data   block_data = 2;
}

...

> // Data contains the set of transactions, evidence, intermediate state roots,

Author: good point https://github.com/celestiaorg/celestia-core/issues/649

// and messages to be included in the block
message Data {
 // Txs that will be applied by state @ block.Height+1.
 // NOTE: not all txs here are valid.  We're just agreeing on the order first.
 // This means that block.AppHash does not include these txs.
 repeated bytes txs = 1;

 IntermediateStateRoots intermediate_state_roots = 2 [(gogoproto.nullable) = false];
 EvidenceList           evidence                 = 3 [(gogoproto.nullable) = false];
 Messages               messages                 = 4 [(gogoproto.nullable) = false];
 uint64                 original_square_size     = 5;
 bytes                  hash                     = 6; // <-- The hash has been added so that the erasure step can be performed in the app
}
```

Here is where the new ABCI++ method is called to perform arbitrary logic in the app before creating the proposal block. `consensus/state.go`

```go

> // CreateProposalBlock calls state.MakeBlock with evidence from the evpool

Reviewer 3: what is `evpool` ?

Author: evidence pool https://github.com/celestiaorg/celestia-core/blob/86b292b620b34a360fe3fc754ec55d0b42bec400/internal/evidence/pool.go#L29-L52

- celestiaorg/cosmos-sdk [#63](https://github.com/celestiaorg/cosmos-sdk/pull/63)
- celestiaorg/celestia-app [#214](https://github.com/celestiaorg/celestia-app/pull/214)

### PrepareProposal [#637](https://github.com/celestiaorg/celestia-core/pull/637) and [#224](https://github.com/celestiaorg/celestia-app/pull/224)

The way that we create proposal blocks will be refactored (previously [`PrePreprocessTxs`](https://github.com/celestiaorg/celestia-app/blob/0363f0410d9d6bf0e51ac92afcaa9be7c0d1ba07/app/abci.go#L17-L108)) to accommodate the new features.

```go
// PrepareProposal separates messages from transactions, malleates those transactions,
// estimates the square size, fills the data square, and calculates the data hash from
// the erasure data
func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
   squareSize := app.estimateSquareSize(req.BlockData)

   dataSquare, data, err := SplitShares(app.txConfig, squareSize, req.BlockData)
   if err != nil {
       panic(err)
   }

   eds, err := da.ExtendShares(squareSize, dataSquare)
   if err != nil {
       // ... log error
       panic(err)
   }

   dah := da.NewDataAvailabilityHeader(eds)

>    data.Hash = dah.Hash() // <-- here we are setting the data hash before we pass the block data back to tendermint

Reviewer: This means we can delete some more of the custom logic in tendermint or will still compute this in core additionally?

Reviewer: Can we clarify what core still has to do still and what is taken over by the app now?

Author: yes, that's a great idea https://github.com/celestiaorg/celestia-app/pull/231/commits/40861a89e557e23c301e772ca8420440dbe366eb

as mentioned in the new paragraph, it does not allow us to remove that much code from tendermint, but it does dramatically increase efficiency as we are no long computing the hash every time we decode the block, and allows us to keep the rest of tendermint that expects this functionality the same (mainly tests).

   data.OriginalSquareSize = squareSize

   return abci.ResponsePrepareProposal{
       BlockData: data,
   }
}
```

> We estimate the square size by assuming that all of the malleable transactions in the block have a valid commitment for whatever square size that we end up picking, and then quickly iterating through the block data to add up the expected lengths of each message/transaction. Please see [here](https://github.com/celestiaorg/celestia-app/blob/e18d8d2301a96702e1bf684735a3620eb059b12f/app/prepare_proposal.go#L47-L130) for more details.

Reviewer 3: ```suggestion
We estimate the square size by assuming that all the malleable transactions in the block have a valid commitment for whatever square size that we end up picking, and then quickly iterating through the block data to add up the expected lengths of each message/transaction. Please see [here](https://github.com/celestiaorg/celestia-app/blob/e18d8d2301a96702e1bf684735a3620eb059b12f/app/prepare_proposal.go#L47-L130) for more details.
```

// written take up.
type ContiguousShareWriter struct {
   shares       []NamespacedShare
   pendingShare NamespacedShare
   namespace    namespace.ID
}
...
// MessageShareWriter lazily merges messages into shares that will eventually be
// included in a data square. It also has methods to help progressively count
// how many shares the messages written take up.
type MessageShareWriter struct {
   shares [][]NamespacedShare
   count  int
}
```

These types are combined in a new celestia-app type, `squareWriter`, which is responsible for atomically writing transactions and their corresponding messages to the data square and the returned block data.

```go
// squareWriter writes a data square using provided block data. It also ensures
// that message and their corresponding txs get written to the square
// atomically.
type squareWriter struct {
   txWriter  *coretypes.ContiguousShareWriter
   msgWriter *coretypes.MessageShareWriter
   ...
}
```

```go
// WriteSquare uses the provided block data to create a flattened data square.
// Any MsgWirePayForMessages are malleated, and their corresponding
// MsgPayForMessage and Message are written atomically. If there are

> // transactions that will node fit in the given square size, then they are

Reviewer: ```suggestion
// transactions that will not fit in the given square size, then they are
```
?

// discarded. This is reflected in the returned block data. Note: pointers to
// block data are only used to avoid dereferencing, not because we need the block
// data to be mutable.

> func WriteSquare(txConf client.TxConfig, squareSize uint64, data *core.Data) ([][]byte, *core.Data, error) {

Reviewer: nit: WriteSquare sounds like this actually writes the square into storage or sth.

Author: agree f7b6fcb7ba324d75f52a484585090c5844ca8a22

           // was not room for this particular tx + msg, there might be room
           // for other txs or even other smaller messages
           continue
       }
       processedTxs = append(processedTxs, malTx)
       messages.MessagesList = append(messages.MessagesList, message)
   }

   sort.Slice(messages.MessagesList, func(i, j int) bool {
       return bytes.Compare(messages.MessagesList[i].NamespaceId, messages.MessagesList[j].NamespaceId) < 0
   })

   return sqwr.export(), &core.Data{
       Txs:                    processedTxs,
       Messages:               messages,
       Evidence:               data.Evidence,
       IntermediateStateRoots: data.IntermediateStateRoots,
   }, nil
}

...

// writeTx marshals the tx and lazily writes it to the square. Returns true if
// the write was successful, false if there was not enough room in the square.
func (sqwr *squareWriter) writeTx(tx []byte) (ok bool, err error) {
   delimTx, err := coretypes.Tx(tx).MarshalDelimited()
   if err != nil {
       return false, err
   }

   if !sqwr.hasRoomForTx(delimTx) {
       return false, nil
   }

   sqwr.txWriter.Write(delimTx)
   return true, nil
}

> // writeMalleated malleates a MsgWirePayForMessage into a MsgPayForMessage and

Reviewer: ```suggestion
// writeMalleatedTx malleates a MsgWirePayForMessage into a MsgPayForMessage and
```

# ADR 001: ABCI++ Adoption

## Changelog

- 03/03/2022: Initial Commit

## Context

Among other things, ABCI++ enables arbitrary logic to be performed by the application to create and verify proposal blocks.

We need this functionality in order for block producers to

- Pick a square size / Fill the data square efficiently
- Malleate user transactions
- Separate rollup block data from the stateful portion of the transaction
- Use the appropriate signature for the selected block size
- Follow the non-interactive default rules (not done here)
- Create celestia specific data hash by erasuring the block data

We also need this functionality for validators to verify that

- Messages that are paid for by users are actually included in the block

> - The block producer doesn’t also include messages that are not paid for

Reviewer: But they would be allowed to include zero fee messages, right?

Author: yes, that's a good point! It would be more accurate to say that it has a corresponding `MsgPayForMessage`, which is not necessarily the same as paying for it

# ADR 001: ABCI++ Adoption

## Changelog

- 2022-03-03: Initial Commit

## Context

Among other things, ABCI++ enables arbitrary logic to be performed by the application to create and verify proposal blocks.

> We need this functionality in order for block producers to

Reviewer 3: ```suggestion
We need this functionality in order for block producers to:
```

- Pick a square size / fill the data square efficiently
- Malleate user transactions
- Separate rollup block data from the stateful portion of the transaction
- Use the appropriate signature for the selected block size
- Follow the [non-interactive default rules](https://github.com/celestiaorg/celestia-specs/blob/65d4d528beafaf8d84f7eb805db940f05f2a2c93/src/rationale/message_block_layout.md#non-interactive-default-rules) (not done here)
- Create Celestia-specific data hash by erasure-coding the block data

> We also need this functionality for validators to verify that

Reviewer 3: ```suggestion
We also need this functionality for validators to verify that:
```

- For every `MsgPayForMessage` included in the block, there is also a corresponding message and vice versa.

> - That the data hash represents the properly-erasure-coded block data for the selected block size.

Reviewer 3: ```suggestion
- The data hash represents the properly-erasure-coded block data for the selected block size.
```

> - Included messages are arranged in the expected locations in the square according to the non-interactive default rules (not done here)

Reviewer 3: ```suggestion
- The included messages are arranged in the expected locations in the square according to the non-interactive default rules (not done here)
```

Technically, we don’t have to use ABCI++ yet, we could still test some needed features in the upcoming testnet without it. However, these implementations would not be representative of the implementations that would actually make it to mainnet, as they would have to be significantly different from their ABCI++ counterparts. The decision to adopt ABCI++ earlier becomes easier considering that the tendermint team has already done most of the heavy lifting, and it is possible to start working on the needed features without waiting for the cosmos-sdk team to use them. We explain our plans below to do just this, by using a subset of ABCI++ (ABCI+?) using only the new methods that are necessary, finished, and easy to incorporate into the cosmos-sdk.

## Alternative Approaches

> While the adoption of ABCI++ is inevitable given the already made decision by upstream to implement it, here are some alternatives to the features that we need that do not use ABCI++.

Reviewer 3: ```suggestion
While the adoption of ABCI++ is inevitable given the already made decision by upstream to implement it, here are some alternatives to the features that we need that do not use ABCI++:
```

> Alternatives for Message Inclusion

Reviewer 3: ```suggestion
- [Alternatives for Message Inclusion.](https://github.com/celestiaorg/celestia-app/blob/92341dd68ee6e555ec6c0bb780afa3a1c8243a93/adrs/adr008:adopt-ABC%2B%2B-early.md#alternative-approaches)
```

> https://github.com/celestiaorg/celestia-app/blob/92341dd68ee6e555ec6c0bb780afa3a1c8243a93/adrs/adr008:adopt-ABC%2B%2B-early.md#alternative-approaches

Reviewer 3: ```suggestion

```

> 

Reviewer 3: ```suggestion
```

> Alternatives for Picking a square size

Reviewer 3: ```suggestion
- [Alternatives for Picking a square size.](https://github.com/celestiaorg/celestia-core/issues/454)
```

> https://github.com/celestiaorg/celestia-core/issues/454

Reviewer 3: ```suggestion
```
